File: packages/sdk-go/mcp_server_test.go
package rateguard

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"strings"
	"testing"
)

// runMCPSession feeds newline-delimited JSON-RPC requests through ServeMCP
// and returns the decoded responses in order.
func runMCPSession(t *testing.T, sdk *SDK, requests ...string) []map[string]any {
	t.Helper()

	input := strings.Join(requests, "\n") + "\n"
	var output bytes.Buffer

	if err := sdk.ServeMCP(context.Background(), strings.NewReader(input), &output); err != nil {
		t.Fatalf("ServeMCP returned error: %v", err)
	}

	var responses []map[string]any
	scanner := bufio.NewScanner(&output)
	for scanner.Scan() {
		var decoded map[string]any
		if err := json.Unmarshal(scanner.Bytes(), &decoded); err != nil {
			t.Fatalf("invalid JSON response %q: %v", scanner.Text(), err)
		}
		responses = append(responses, decoded)
	}
	return responses
}

func TestMCPServerHandshakeAndToolsList(t *testing.T) {
	sdk := New(Config{Preset: "dev"})

	responses := runMCPSession(t, sdk,
		`{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"protocolVersion":"2025-06-18","capabilities":{},"clientInfo":{"name":"test","version":"0"}}}`,
		`{"jsonrpc":"2.0","method":"notifications/initialized"}`,
		`{"jsonrpc":"2.0","id":2,"method":"tools/list"}`,
	)

	if len(responses) != 2 {
		t.Fatalf("expected 2 responses (initialize, tools/list), got %d: %v", len(responses), responses)
	}

	initResult, ok := responses[0]["result"].(map[string]any)
	if !ok {
		t.Fatalf("initialize response missing result: %v", responses[0])
	}
	if initResult["protocolVersion"] != mcpProtocolVersion {
		t.Errorf("protocolVersion = %v, want %s", initResult["protocolVersion"], mcpProtocolVersion)
	}

	listResult, ok := responses[1]["result"].(map[string]any)
	if !ok {
		t.Fatalf("tools/list response missing result: %v", responses[1])
	}
	tools, ok := listResult["tools"].([]any)
	if !ok || len(tools) != 7 {
		t.Fatalf("tools/list should return 7 tools, got %v", listResult["tools"])
	}
}

func TestMCPServerToolsCall(t *testing.T) {
	sdk := New(Config{Preset: "dev"})

	responses := runMCPSession(t, sdk,
		`{"jsonrpc":"2.0","id":7,"method":"tools/call","params":{"name":"get_rate_limit_state","arguments":{"key":"agent-1"}}}`,
	)

	if len(responses) != 1 {
		t.Fatalf("expected 1 response, got %d", len(responses))
	}
	result, ok := responses[0]["result"].(map[string]any)
	if !ok {
		t.Fatalf("tools/call missing result: %v", responses[0])
	}
	if isError, _ := result["isError"].(bool); isError {
		t.Fatalf("tools/call reported error: %v", result)
	}
	content, ok := result["content"].([]any)
	if !ok || len(content) != 1 {
		t.Fatalf("expected one content block, got %v", result["content"])
	}
}

func TestMCPServerUnknownMethod(t *testing.T) {
	sdk := New(Config{Preset: "dev"})

	responses := runMCPSession(t, sdk,
		`{"jsonrpc":"2.0","id":3,"method":"resources/list"}`,
	)

	if len(responses) != 1 {
		t.Fatalf("expected 1 response, got %d", len(responses))
	}
	errObj, ok := responses[0]["error"].(map[string]any)
	if !ok {
		t.Fatalf("expected error response, got %v", responses[0])
	}
	if code, _ := errObj["code"].(float64); int(code) != jsonrpcMethodNotFound {
		t.Errorf("error code = %v, want %d", errObj["code"], jsonrpcMethodNotFound)
	}
}

func TestMCPServerToolErrorInBand(t *testing.T) {
	sdk := New(Config{Preset: "dev"})

	// Missing required "key" argument → in-band isError, not a JSON-RPC error.
	responses := runMCPSession(t, sdk,
		`{"jsonrpc":"2.0","id":4,"method":"tools/call","params":{"name":"get_rate_limit_state","arguments":{}}}`,
	)

	result, ok := responses[0]["result"].(map[string]any)
	if !ok {
		t.Fatalf("expected in-band tool error, got %v", responses[0])
	}
	if isError, _ := result["isError"].(bool); !isError {
		t.Fatalf("expected isError=true, got %v", result)
	}
}

File: packages/sdk-node/test/mcp-server.test.ts
import { describe, it, expect } from 'vitest';
import { Readable, Writable } from 'node:stream';
import { RateGuard, serveMCP, MCP_PROTOCOL_VERSION } from '../src/index.js';

interface JSONRPCResponse {
  jsonrpc: string;
  id?: unknown;
  result?: Record<string, unknown>;
  error?: { code: number; message: string };
}

/** Pipes newline-delimited JSON-RPC lines through serveMCP and returns the parsed responses. */
async function drive(guard: RateGuard, lines: string[]): Promise<JSONRPCResponse[]> {
  const chunks: string[] = [];
  const stdout = new Writable({
    write(chunk, _encoding, callback) {
      chunks.push(String(chunk));
      callback();
    },
  });
  const stdin = Readable.from(lines.map((line) => line + '\n'));

  await serveMCP(guard, stdin, stdout);

  return chunks
    .join('')
    .split('\n')
    .filter((line) => line.trim().length > 0)
    .map((line) => JSON.parse(line) as JSONRPCResponse);
}

describe('MCP stdio server', () => {
  it('serves initialize, tools/list, and tools/call over newline-delimited JSON-RPC', async () => {
    const guard = new RateGuard({ preset: 'dev' });
    const responses = await drive(guard, [
      JSON.stringify({ jsonrpc: '2.0', id: 1, method: 'initialize', params: {} }),
      JSON.stringify({ jsonrpc: '2.0', method: 'notifications/initialized' }),
      JSON.stringify({ jsonrpc: '2.0', id: 2, method: 'ping' }),
      JSON.stringify({ jsonrpc: '2.0', id: 3, method: 'tools/list' }),
      JSON.stringify({
        jsonrpc: '2.0',
        id: 4,
        method: 'tools/call',
        params: { name: 'get_rate_limit_state', arguments: { key: 'agent-1' } },
      }),
    ]);

    // The notification must not produce a response line.
    expect(responses).toHaveLength(4);

    const init = responses.find((r) => r.id === 1)!;
    expect(init.result).toMatchObject({
      protocolVersion: MCP_PROTOCOL_VERSION,
      capabilities: { tools: {} },
      serverInfo: { name: 'rateguard' },
    });

    const ping = responses.find((r) => r.id === 2)!;
    expect(ping.result).toEqual({});

    const list = responses.find((r) => r.id === 3)!;
    const listed = (list.result!.tools as Array<{ name: string }>).map((tool) => tool.name);
    expect(listed).toEqual(guard.mcpTools().map((tool) => tool.name));

    const call = responses.find((r) => r.id === 4)!;
    expect(call.result!.isError).toBe(false);
    const content = call.result!.content as Array<{ type: string; text: string }>;
    const state = JSON.parse(content[0]!.text) as Record<string, unknown>;
    expect(state.key).toBe('agent-1');
    expect(typeof state.allowed).toBe('boolean');
    // Pre-flight peek semantics: querying must not consume a token.
    expect(state.remaining).toBe(guard.runtime.config.rateLimit.burst);
  });

  it('returns -32601 for unknown methods but stays silent for unknown notifications', async () => {
    const guard = new RateGuard({ preset: 'dev' });
    const responses = await drive(guard, [
      JSON.stringify({ jsonrpc: '2.0', id: 7, method: 'resources/list' }),
      JSON.stringify({ jsonrpc: '2.0', method: 'notifications/nonexistent' }),
    ]);

    expect(responses).toHaveLength(1);
    expect(responses[0]!.id).toBe(7);
    expect(responses[0]!.error?.code).toBe(-32601);
    expect(responses[0]!.error?.message).toContain('resources/list');
  });

  it('returns -32700 for unparseable lines and keeps serving afterwards', async () => {
    const guard = new RateGuard({ preset: 'dev' });
    const responses = await drive(guard, [
      'this is not json {',
      JSON.stringify({ jsonrpc: '2.0', id: 8, method: 'ping' }),
    ]);

    expect(responses).toHaveLength(2);
    expect(responses[0]!.error?.code).toBe(-32700);
    expect(responses[0]!.id).toBeUndefined();
    expect(responses[1]!.id).toBe(8);
    expect(responses[1]!.result).toEqual({});
  });

  it('returns -32602 when tools/call is missing params.name', async () => {
    const guard = new RateGuard({ preset: 'dev' });
    const responses = await drive(guard, [
      JSON.stringify({ jsonrpc: '2.0', id: 9, method: 'tools/call', params: {} }),
      JSON.stringify({ jsonrpc: '2.0', id: 10, method: 'tools/call', params: { name: '' } }),
    ]);

    expect(responses).toHaveLength(2);
    for (const response of responses) {
      expect(response.error?.code).toBe(-32602);
      expect(response.error?.message).toContain('params.name');
    }
  });

  it('reports tool-level failures in-band with isError, not as protocol errors', async () => {
    const guard = new RateGuard({ preset: 'dev' });
    const responses = await drive(guard, [
      JSON.stringify({ jsonrpc: '2.0', id: 11, method: 'tools/call', params: { name: 'no_such_tool' } }),
      JSON.stringify({
        jsonrpc: '2.0',
        id: 12,
        method: 'tools/call',
        // get_rate_limit_state without a key makes the tool handler throw.
        params: { name: 'get_rate_limit_state', arguments: {} },
      }),
    ]);

    expect(responses).toHaveLength(2);
    for (const response of responses) {
      expect(response.error).toBeUndefined();
      expect(response.result!.isError).toBe(true);
      const content = response.result!.content as Array<{ type: string; text: string }>;
      expect(content[0]!.type).toBe('text');
      expect(content[0]!.text.length).toBeGreaterThan(0);
    }
  });
});
